### ISSUE TITLE
Grid editing quote: some fields on the grid should be locked for editing

### ISSUE THREAD

**Issue by githubbob42** [githubbob42/mingle2github2#3761]
**[Mingle Card: 4085](https://liquidframeworks.mingle.thoughtworks.com/projects/alpine_mobile/cards/4085)**
Steps to reproduce
------------------

```
| | |
|-|-|
|**Version #**|
|3215|
|**OS**|
|windows|
|**Browser**|
|chrome|
|**Username**|
|npfield@soasta1.com|
|**Password**|
|computer45|
  
 

```

1.  Login to Mobile
2.  Select any quote that is not in read only status and go to quote items
3.  Enable grid editing

Expected result
---------------

_Fields such as_ _Price Book item_, Record Type, Catalog UOM should be locked for editing.

_Update: Pricebook item should (and quote) should not be displayed at all. See: #4024_

Actual Result
-------------

_Those fields available for editing but clicking on those fields result in endless spin._

Related Cards
-------------

Please link this to the similar card for locking fields on the declarative screen

#4024

Analysis
--------

The fetch process for ticket items overrides certain fields to always be locked if they exist. This set includes:

SyncID\_\_c  
PBI\_Included\_Quantity\_\_c  
PBI\_Locked\_Description\_\_c  
PBI\_Locked\_Discount\_\_c  
PBI\_Locked\_Price\_Markup\_\_c  
PBI\_Maximum\_Quantity\_\_c  
PBI\_Minimum\_Quantity\_\_c  
PBI\_Minimum\_Charge\_\_c  
PBI\_Pricing\_Method\_\_c  
PBI\_Required\_Item\_\_c  
PBI\_Requires\_Parent\_Item\_\_c  
PBI\_Surcharge\_Env\_\_c  
PBI\_Surcharge\_Fuel\_\_c  
Breadcrumb\_\_c  
Surcharge\_Rate\_Env\_\_c  
Surcharge\_Rate\_Fuel\_\_c  
Catalog\_Track\_Inventory\_\_c  
Item\_Code\_\_c  
Catalog\_Record\_Type\_\_c  
PBI\_Default\_Quantity\_\_c  
Catalog\_UOM\_\_c

Additionally, a bug on the quote items screen (#4140) is causing the grid to not correctly honor the layout (so record type is editable).

UPDATE: BR – 6/6/16: Talking with DP about another card, he has implemented 2 new features to the “layout/form-edit” bindings (one is already merged, the other should go in hopefully soon).  He added skipFields and lockFields;

*   skipFields does not display the field on the form at all, while
*   getLockedFields simply makes the “PBI\_Locked\_\*” fields readonly (not merged yet, [https://github.com/dpitre/alpine-mobile/tree/4116\_incremental\_ticket\_items](https://github.com/dpitre/alpine-mobile/tree/4116_incremental_ticket_items))

We can either wait for DP’s branch/PR to merge (+1 week after 4100 is merged) or we can implement it as part of this card. (TOM SAYS TO WAIT)

Looks like once DP’s PR is merged, we’ll need to add a test for readonlyFields to quote/item/grid/row-edit.js:

```
<!-- ko let: {
  column: $editLayoutItem || $detailLayoutItem,
  $isEditable: $editLayoutItem && $editLayoutItem.editableForUpdate **&& !~$list.readonlyFields.indexOf($primaryComponent.details.name)**,
  $css: {}
} →
```

UPDATE: BR – 6/242016: 

Actually, we are trying to move away from the above method of modifying the layout/metatadata and move lock field management to an override filter.  DP added an override filter for ticket items as part of card \[alpine\_mobile/#4191\] Simplify Locking Fields for Ticket Items.  

We should leverage a similar approach for quote items.

Test Plan
---------

Not sure what happened but it looks like the last PR build did not include my changes.  Rebuilt and retested on hunterpeak and all looks good...  please test again.

### Checklist for Merging

*   Enter the Pull Request number in the card properties
*   How is this tested in Selenium? 
*   Do we have adequate unit testing? 
*   Where is this covered in documentation?
*   Where is this in Tarantula?

**Comment by githubbob42** [comment]
Matthew Todd : 07/05/2016 17:46 CDT
Failed PR Build: 4.0.1.pr1889 Admin User: ajain@tarantula3.com Field User: mattfield@test3.com FieldFX URL: https://dev.fieldfx.com Hardware: Dell Latitude E6530 Browser: Chrome Test Case: TBD Quote: Dynamic Parent in Console (Q-DB1-0C1-2BDE) This has failed testing on the 4.0.1.pr1889 build. For some reason, some of the rows in the Catalog UOM column are still editable. However, clicking on them does not result in an endless spin. See attached screenshot for more information.

**Comment by githubbob42** [comment]
Bob Roth : 07/12/2016 13:28 CDT
Not sure what happened but it looks like the PR build did not include my changes. Rebuilt and retested on hunterpeak and all looks good... please test again

**Comment by githubbob42** [comment]
Matthew Todd : 07/20/2016 09:22 CDT
Passed PR Build: 4.0.1.pr1889 Field User: qafield@test4.com FieldFX URL: https://dev.fieldfx.com Hardware: Dell Latitude E6530 Browser: Chrome 51.0.2704.106 m Test Case: Inline Editing of Quote Items (4085) This has passed testing with the 4.0.1.pr1889 branch. The previously unlocked fields are now locked for editing. I also updated the ‘Inline Editing of Quote Items’ test case to include the list of fields that should be locked.

**Comment by githubbob42** [comment]
**Pull Request #1889**
